<a href="https://colab.research.google.com/github/cycywanyeki20/Movies-Phase1-Project/blob/master/Phase_2_Code_Challenge_Notebook.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Phase 2 Code Challenge

This code challenge is designed to test your understanding of the Phase 2 material. It covers:

- Normal Distribution
- Statistical Tests
- Bayesian Statistics
- Linear Regression

_Read the instructions carefully_. You will be asked both to write code and to answer short answer questions.

## Code Tests

We have provided some code tests for you to run to check that your work meets the item specifications. Passing these tests does not necessarily mean that you have gotten the item correct - there are additional hidden tests. However, if any of the tests do not pass, this tells you that your code is incorrect and needs changes to meet the specification. To determine what the issue is, read the comments in the code test cells, the error message you receive, and the item instructions.

## Short Answer Questions

For the short answer questions...

* _Use your own words_. It is OK to refer to outside resources when crafting your response, but _do not copy text from another source_.

* _Communicate clearly_. We are not grading your writing skills, but you can only receive full credit if your teacher is able to fully understand your response.

* _Be concise_. You should be able to answer most short answer questions in a sentence or two. Writing unnecessarily long answers increases the risk of you being unclear or saying something incorrect.

In [50]:
# Run this cell without changes to import the necessary libraries

import itertools
import numpy as np
import pandas as pd
from numbers import Number
from scipy import stats
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

import pickle

---
## Part 1: Normal Distribution [Suggested time: 20 minutes]
---
In this part, you will analyze check totals at a TexMex restaurant. We know that the population distribution of check totals for the TexMex restaurant is normally distributed with a mean of \\$20 and a standard deviation of \\$3.

### 1.1) Create a numeric variable `z_score_26` containing the z-score for a \\$26 check.

**Starter Code**

    z_score_26 =

In [51]:
# YOUR CODE HERE
#raise NotImplementedError()
X = 26  # The value to standardize
mean = 20  # Mean of the population distribution
std_dev = 3  # Standard deviation of the population distribution

# Calculate the z-score
z_score_26 = (X - mean) / std_dev

# Print the z-score
print(z_score_26)

2.0


In [52]:
# This test confirms that you have created a numeric variable named z_score_26

assert isinstance(z_score_26, Number)


### 1.2) Create a numeric variable `p_under_26` containing the approximate proportion of all checks that are less than \\$26.

Hint: Use the answer from the previous question along with the empirical rule, a Python function, or this [z-table](https://www.math.arizona.edu/~rsims/ma464/standardnormaltable.pdf).

**Starter Code**

    p_under_26 =

In [53]:
# YOUR CODE HERE
#raise NotImplementedError()
import scipy.stats as stats

# Define the z-score calculated in the above question
z_score_26 = 2.0

# Calculate the approximate proportion of checks less than $26
p_under_26 = stats.norm.cdf(z_score_26)

# Print the result
print( p_under_26) # The approximation of checks under $26.

0.9772498680518208


we use the stats.norm.cdf(z_score_26) function to calculate the cumulative distribution function of the standard normal distribution, which gives us the proportion of checks less than $26. replace z_score_26 with 2.

In [54]:
# This test confirms that you have created a numeric variable named p_under_26

assert isinstance(p_under_26, Number)

# These tests confirm that p_under_26 is a value between 0 and 1

assert p_under_26 >= 0
assert p_under_26 <= 1


### 1.3) Create numeric variables `conf_low` and `conf_high` containing the lower and upper bounds (respectively) of a 95% confidence interval for the mean of one waiter's check amounts using the information below.

One week, a waiter gets 100 checks with a mean of \\$19 and a standard deviation of \\$3.

**Starter Code**

    n = 100
    mean = 19
    std = 3
    
    conf_low =
    conf_high =

In [55]:
# YOUR CODE HERE
#raise NotImplementedError()
n = 100
mean = 19
std = 3
confidence_level = 0.95  # 95% confidence level
z_score = 1.96  # Z-Score for a 95% confidence interval is typically 1.96 in z-score table

# Calculate the standard error
standard_error = std / (n ** 0.5)
#std: The standard deviation of the sample.
#n: The sample size.
#n ** 0.5: The square root of the sample size.

# Calculate the margin of error
error_margin = z_score * standard_error

# Calculate the lower and upper bounds of the confidence interval
conf_low = mean - error_margin
conf_high = mean + error_margin

# Print the results
print(conf_low) #Lower bound of 95% confidence interval
print(conf_high) #Upper bound of 95% confidence interval



18.412
19.588


To calculate the Z-Score for a 95% confidence interval, you'll use the Z-Score associated with a 95% confidence level, which is typically 1.96.

In [56]:
# These tests confirm that you have created numeric variables named conf_low and conf_high

assert isinstance(conf_low, Number)
assert isinstance(conf_high, Number)

# This test confirms that conf_low is below conf_high

assert conf_low < conf_high

# These statements print your answers for reference to help answer the next question

print('The lower bound of the 95% confidence interval is {}'.format(conf_low))
print('The upper bound of the 95% confidence interval is {}'.format(conf_high))


The lower bound of the 95% confidence interval is 18.412
The upper bound of the 95% confidence interval is 19.588


### 1.4) Short Answer: Interpret the 95% confidence interval you just calculated in Question 1.3.

The 95% interval means that we are 95% confident that the true mean of one waiter's check amounts at the restaurant falls between this range of $ 18.41 and $ 19. 59 dollars

---
## Part 2: Statistical Testing [Suggested time: 20 minutes]
---
The TexMex restaurant recently introduced queso to its menu.

We have a random sample containing 2000 check totals, all from different customers: 1000 check totals for orders without queso ("no queso") and 1000 check totals for orders with queso ("queso").

In the cell below, we load the sample data for you into the arrays `no_queso` and `queso` for the "no queso" and "queso" order check totals, respectively.

In [57]:
# Run this cell without changes

# Load the sample data
no_queso = pickle.load(open('./no_queso.pkl', 'rb'))
queso = pickle.load(open('./queso.pkl', 'rb'))

### 2.1) Short Answer: State null and alternative hypotheses to use for testing whether customers who order queso spend different amounts of money from customers who do not order queso.

H0 - There is no much difference between the amount spent by customers who order queso and those who do not order queso.

H1 - The customers that order queso spend a different amount from the customers that do not order queso.

### 2.2) Short Answer: What would it mean to make a Type I error for this specific hypothesis test?

Your answer should be _specific to this context,_  not a general statement of what Type I error is.

A type one error in this case would mean rejecting the null hypothesis which could be true. The conclusion made would mean that there is indeed a difference in the amounts spent by queso customers and non-queso customers when in fact the difference is very small. Such an error will lead to changes in prices and wrong marketing strategies.

### 2.3) Create a numeric variable `p_value` containing the p-value associated with a statistical test of your hypotheses.

You must identify and implement the correct statistical test for this scenario. You can assume the two samples have equal variances.

Hint: Use `scipy.stats` to calculate the answer - it has already been imported as `stats`. Relevant documentation can be found [here](https://docs.scipy.org/doc/scipy/reference/stats.html#statistical-tests).

**Starter Code**

    p_value =

In [58]:
# YOUR CODE HERE
#raise NotImplementedError()
import scipy.stats as stats

# State the two samples (no_queso and queso)
no_queso = [1000] #orders without queso
queso = [1000]  #orders with queso

# Perform a two-sample t-test while assuming variances are equal
t_statistic, p_value = stats.ttest_ind(no_queso, queso, equal_var=True)

# Print the p-value
print("P-Value:", p_value)

P-Value: nan


t_statistic: This variable contains the calculated t-statistic, which measures the difference in means between the two samples (orders with and without queso) and compares it to the expected difference under the null hypothesis.

p_value: This variable contains the p-value associated with the t-test. The p-value indicates the probability of observing the data, assuming that the null hypothesis is true. A smaller p-value suggests stronger evidence against the null hypothesis.

In [59]:
# These tests confirm that you have created a numeric variable named p_value

assert isinstance(p_value, Number)


### 2.4) Short Answer: Can you reject the null hypothesis using a significance level of $\alpha$ = 0.05? Explain why or why not.

No, rejecting the null hypothesis is only valid if the calculated p valued is less that a=0.05.This suggests a significant difference in the spending habits of the two groups of customers. Otherwise, the null hypothesis stands.

---
## Part 3: Bayesian Statistics [Suggested time: 15 minutes]
---
A medical test is designed to diagnose a certain disease. The test has a false positive rate of 10%, meaning that 10% of people without the disease will get a positive test result. The test has a false negative rate of 2%, meaning that 2% of people with the disease will get a negative result. Only 1% of the population has this disease.

### 3.1) Create a numeric variable `p_pos_test` containing the probability of a person receiving a positive test result.

Assume that the person being tested is randomly selected from the broader population.

**Starter Code**
    
    false_pos_rate = 0.1
    false_neg_rate = 0.02
    population_rate = 0.01
    
    p_pos_test =

In [60]:
# YOUR CODE HERE
#raise NotImplementedError()
#Use conditional probability
false_positive_rate = 0.1  # A person does not have the disease and tests positive
false_negative_rate = 0.02  # A person has the disease but tests negative
population_rate = 0.01  # Rate of people with the disease

# Probability of a false negative rate
positive_with_disease = 1 - false_negative_rate

# Probability of false positive rate
positive_without_disease = false_positive_rate

# Probability of a random person having the disease
random_disease = population_rate

# Calculate the overall probability of testing positive using the law of total probability
p_pos_test = (positive_with_disease * random_disease) + (positive_without_disease * (1 - random_disease))

# Print the result
print(p_pos_test) # Overall probability of testing positive


0.10880000000000001


In [61]:
# This test confirms that you have created a numeric variable named p_pos_test

assert isinstance(p_pos_test, Number)

# These tests confirm that p_pos_test is a value between 0 and 1

assert p_pos_test >= 0
assert p_pos_test <= 1


### 3.2) Create a numeric variable `p_pos_pos` containing the probability of a person actually having the disease if they receive a positive test result.

Assume that the person being tested is randomly selected from the broader population.

Hint: Use your answer to the previous question to help answer this one.

**Starter Code**
    
    false_pos_rate = 0.1
    false_neg_rate = 0.02
    population_rate = 0.01
    
    p_pos_pos =

In [62]:
# YOUR CODE HERE
#raise NotImplementedError()
# Using the values from the previous question
positive_with_disease = 1 - false_negative_rate
random_disease = population_rate

# Calculate the probability of a person actually having the disease if they receive a positive test result
p_pos_pos = (positive_with_disease * random_disease) / p_pos_test

# Print the result
print("Probability of having the disease if a positive test result is received:", p_pos_pos)

Probability of having the disease if a positive test result is received: 0.09007352941176469


In [63]:
# This test confirms that you have created a numeric variable named p_pos_pos

assert isinstance(p_pos_pos, Number)

# These tests confirm that p_pos_pos is a value between 0 and 1

assert p_pos_pos >= 0
assert p_pos_pos <= 1


---

## Part 4: Linear Regression [Suggested Time: 20 min]
---

In this section, you'll run regression models with advertising data. The dataset includes the advertising spending and sales for 200 products. These products are a random sample of 200 products from a larger population of products.

The dataset has three features - `TV`, `radio`, and `newspaper` - that describe how many thousands of advertising dollars were spent promoting the product. The target, `sales`, describes how many millions of dollars in sales the product had.

In [64]:
# Run this cell without changes

import statsmodels
from statsmodels.formula.api import ols

data = pd.read_csv('./advertising.csv').drop('Unnamed: 0', axis=1)
X = data.drop('sales', axis=1)
y = data['sales']

### 4.1) Create a variable `tv_mod_summ` containing the summary of a fit StatsModels `ols` model for a linear regression using `TV` to predict `sales`.

**Starter Code**

    tv_mod = ols(
    tv_mod_summ = tv_mod.summary()

In [65]:
# YOUR CODE HERE
#raise NotImplementedError()
import statsmodels.api as sm

#  DataFrame named Advertising data with 'TV' and 'sales' columns
X = data['TV']  # TV advertising spending
X = sm.add_constant(X)  # intercept
y = data['sales']  # Sales

# Fit the linear regression model
model = sm.OLS(y, X).fit()

# Get the summary of the model
tv_mod_summ = model.summary()

# Print the summary
print(tv_mod_summ)

                            OLS Regression Results                            
Dep. Variable:                  sales   R-squared:                       0.612
Model:                            OLS   Adj. R-squared:                  0.610
Method:                 Least Squares   F-statistic:                     312.1
Date:                Tue, 24 Oct 2023   Prob (F-statistic):           1.47e-42
Time:                        09:59:10   Log-Likelihood:                -519.05
No. Observations:                 200   AIC:                             1042.
Df Residuals:                     198   BIC:                             1049.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          7.0326      0.458     15.360      0.0

In [66]:
# This test confirms that you have created a variable named tv_mod_summ containing a StatsModels Summary

assert type(tv_mod_summ) == statsmodels.iolib.summary.Summary

# This statement prints your answer for reference to help answer the next question

print(tv_mod_summ)


                            OLS Regression Results                            
Dep. Variable:                  sales   R-squared:                       0.612
Model:                            OLS   Adj. R-squared:                  0.610
Method:                 Least Squares   F-statistic:                     312.1
Date:                Tue, 24 Oct 2023   Prob (F-statistic):           1.47e-42
Time:                        09:59:10   Log-Likelihood:                -519.05
No. Observations:                 200   AIC:                             1042.
Df Residuals:                     198   BIC:                             1049.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          7.0326      0.458     15.360      0.0

### 4.2) Short Answer: Is there a statistically significant relationship between TV advertising spend and sales in this model? How did you determine this from the model output?

This question is asking you to use your findings from the sample in your dataset to make an inference about the relationship between TV advertising spend and sales in the broader population.

Yes there is a statistically significant relation between TV advertising spend and sales in this models because the p value is less than 0.05.

### 4.3) Short Answer: Run the cell below to produce a correlation matrix. Given the output, would you expect any collinearity issues if you included all of these features in one regression model?

In [67]:
# Run this cell

X.corr()

const   TV
const    NaN  NaN
TV       NaN  1.0

Yes, I would expect collinearity issues because of the high correlation. Values close to 1 leading to multicolinearity which is a problem in regression analysis.

### 4.4) Create a variable `all_mod_summ` containing the summary of a fit StatsModels `ols` model for a multiple regression using `TV`, `radio`, and `newspaper` to predict `sales`.

**Starter Code**

    all_mod = ols(
    all_mod_summ = all_mod.summary()

In [68]:
# YOUR CODE HERE
#raise NotImplementedError()
import statsmodels.api as sm

X = data[['TV', 'radio', 'newspaper']]
X = sm.add_constant(X)  # intercept
y = data['sales']  # Sales

# Fit the multiple regression model
model = sm.OLS(y, X).fit()

# Get the summary of the model
all_mod_summ = model.summary()

# Print the summary
print(all_mod_summ)

                            OLS Regression Results                            
Dep. Variable:                  sales   R-squared:                       0.897
Model:                            OLS   Adj. R-squared:                  0.896
Method:                 Least Squares   F-statistic:                     570.3
Date:                Tue, 24 Oct 2023   Prob (F-statistic):           1.58e-96
Time:                        09:59:10   Log-Likelihood:                -386.18
No. Observations:                 200   AIC:                             780.4
Df Residuals:                     196   BIC:                             793.6
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          2.9389      0.312      9.422      0.0

In [69]:
# This test confirms that you have created a variable named all_mod_summ containing a StatsModels Summary

assert type(all_mod_summ) == statsmodels.iolib.summary.Summary

# This statement prints your answer for reference to help answer the next question

print(all_mod_summ)


                            OLS Regression Results                            
Dep. Variable:                  sales   R-squared:                       0.897
Model:                            OLS   Adj. R-squared:                  0.896
Method:                 Least Squares   F-statistic:                     570.3
Date:                Tue, 24 Oct 2023   Prob (F-statistic):           1.58e-96
Time:                        09:59:10   Log-Likelihood:                -386.18
No. Observations:                 200   AIC:                             780.4
Df Residuals:                     196   BIC:                             793.6
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          2.9389      0.312      9.422      0.0

### 4.5) Short Answer: Does this model do a better job of explaining sales than the previous model using only the `TV` feature? Explain how you determined this based on the model output.

Yes the models does a better job than the previous one. This is because the r2 value of this model is higher (0.897) compared to the previous values of ( 0.612)